designs/SpanMarshallers.md

```diff
@@ -0,0 +1,262 @@
+
+An alternative option to fully inlining the stub would be to extend the model described in the [Struct Marshalling design doc](./StructMarshalling.md) to have custom support for collection-like types. By extending the model to be built with generic collection types in mind, many of the cons of the first approach would be resolved.
+
+Span marshalling would still be implemented with similar semantics as mentioned above in the Empty Spans section. Additional marshallers would still be provided as mentioned in the Additional proposed in-source marshallers section, but the non-`null` span marshaller and the no-alloc span marshaller would be able to be used in all cases, not just for empty spans.
+
+### Proposed extension to the custom type marshalling design
+
+Introduce a new attribute named `GenericCollectionMarshallerAttribute`. This attribute would have the following shape:
+
+```csharp
+namespace System.Runtime.InteropServices
+{ 
+    [AttributeUsage(AttributeTargets.Struct | AttributeTargets.Class)]
+    public sealed class GenericCollectionMarshallerAttribute : Attribute
+    {
+        public GenericCollectionMarshallerAttribute();
+    }
+}
+```
+
+The attribute would be used with a collection type like `Span<T>` as follows:
+
+```csharp
+[NativeTypeMarshalling(typeof(DefaultSpanMarshaler<>))]
+public ref struct Span<T>
+{
+  ...
+}
+
+[GenericCollectionMarshaller]
+public ref struct DefaultSpanMarshaler<T>
+{
+
+}
+```
+
+The `GenericCollectionMarshallerAttribute` attribute is applied to a generic marshaler type with the "collection marshaller" shape described below. Since generic parameters cannot be used in attributes, open generic types will be permitted in the `NativeTypeMarshallingAttribute` constructor as long as they have the same arity as the type the attribute is applied to and generic parameters provided to the applied-to type can also be used to construct the type passed as a parameter.
+
+#### Generic collection marshaller shape
+
+A generic collection marshaller would be required to have the following shape, in addition to the requirements for marshaler types used with the `NativeTypeMarshallingAttribute`, excluding the constructors.
+
+```csharp
+[GenericCollectionMarshaller]
+public struct GenericCollectionMarshallerImpl<T, U, V,...>
+{
+    // these constructors are required if marshalling from managed to native is supported.
+    public GenericCollectionMarshallerImpl(GenericCollection<T, U, V, ...> collection, int nativeSizeOfElements);
+    public GenericCollectionMarshallerImpl(GenericCollection<T, U, V, ...> collection, Span<byte> stackSpace, int nativeSizeOfElements); // optional
+    
+    public const int StackBufferSize = /* */; // required if the span-based constructor is supplied.
+
+    public Span<TCollectionElement> ManagedValues { get; }
+
+    // Required only when unmarshalling is supported.
+    public void SetUnmarshalledCollectionLength(int length);
```

Comment:
Let's fully document each of these members as if we were proposing the API.

Reply:
I see that the `SetUnmarshalledCollectionLength()` would be passed the unmarshal length that is returned by the native call. How does this get mapped on the actual P/Invoke?

Reply:
This would be mapped similarly as the SizeParamIndex and SizeConst fields on MarshalAsAttribute (fetching them immediately after the call).